core/src/core/compact_block.rs

```diff
@@ -0,0 +1,161 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Compact Blocks.
+
+use consensus::VerifySortOrder;
+use core::block::{BlockHeader, Error};
+use core::{Output, ShortId, TxKernel};
+use ser::{self, read_and_verify_sorted, Readable, Reader, Writeable, Writer};
+
+/// Container for full (full) outputs and kernels and kern_ids for a compact block.
+#[derive(Debug, Clone)]
+pub struct CompactBlockBody {
+	/// List of full outputs - specifically the coinbase output(s)
+	pub out_full: Vec<Output>,
+	/// List of full kernels - specifically the coinbase kernel(s)
+	pub kern_full: Vec<TxKernel>,
+	/// List of transaction kernels, excluding those in the full list
+	/// (short_ids)
+	pub kern_ids: Vec<ShortId>,
+}
+
+impl CompactBlockBody {
+	fn validate(&self) -> Result<(), Error> {
+		self.verify_sorted()?;
+		Ok(())
+	}
+
+	// Verify everything is sorted in lexicographical order.
+	fn verify_sorted(&self) -> Result<(), Error> {
+		self.out_full.verify_sort_order()?;
+		self.kern_full.verify_sort_order()?;
+		self.kern_ids.verify_sort_order()?;
+		Ok(())
+	}
+}
+
+impl Readable for CompactBlockBody {
+	fn read(reader: &mut Reader) -> Result<CompactBlockBody, ser::Error> {
+		let (out_full_len, kern_full_len, kern_id_len) =
+			ser_multiread!(reader, read_u64, read_u64, read_u64);
+
+		let out_full = read_and_verify_sorted(reader, out_full_len as u64)?;
+		let kern_full = read_and_verify_sorted(reader, kern_full_len as u64)?;
+		let kern_ids = read_and_verify_sorted(reader, kern_id_len)?;
+
+		Ok(CompactBlockBody {
+			out_full,
+			kern_full,
+			kern_ids,
+		})
+	}
+}
+
+impl Writeable for CompactBlockBody {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ser::Error> {
+		ser_multiwrite!(
+			writer,
+			[write_u64, self.out_full.len() as u64],
+			[write_u64, self.kern_full.len() as u64],
+			[write_u64, self.kern_ids.len() as u64]
+		);
+
+		self.out_full.write(writer)?;
+		self.kern_full.write(writer)?;
+		self.kern_ids.write(writer)?;
+
+		Ok(())
+	}
+}
+
+impl Into<CompactBlockBody> for CompactBlock {
+	fn into(self) -> CompactBlockBody {
+		self.body
+	}
+}
+
+/// Compact representation of a full block.
+/// Each input/output/kernel is represented as a short_id.
+/// A node is reasonably likely to have already seen all tx data (tx broadcast
+/// before block) and can go request missing tx data from peers if necessary to
+/// hydrate a compact block into a full block.
+#[derive(Debug, Clone)]
+pub struct CompactBlock {
+	/// The header with metadata and commitments to the rest of the data
+	pub header: BlockHeader,
+	/// Nonce for connection specific short_ids
+	pub nonce: u64,
+	/// Container for out_full, kern_full and kern_ids in the compact block.
+	pub body: CompactBlockBody,
```

Comment:
do we need it to be `pub`? It's handy, but may be confusing to have methods below and direct access to body